freqtrade/data/dataprovider.py

```diff
@@ -63,9 +79,99 @@ def _set_cached_df(
+        """
+        Set the pairs received to later be used.
+        This only supports 1 Producer right now.
+
+        :param pairlist: List of pairs
+        """
+        self.__producer_pairs[producer_name] = pairlist.copy()
+
+    def get_producer_pairs(self, producer_name: str = "default") -> List[str]:
+        """
+        Get the pairs cached from the producer
+
+        :returns: List of pairs
+        """
+        return self.__producer_pairs.get(producer_name, [])
+
+    def emit_df(
+        self,
+        pair_key: PairWithTimeframe,
+        dataframe: DataFrame
+    ) -> None:
+        """
+        Send this dataframe as an ANALYZED_DF message to RPC
+
+        :param pair_key: PairWithTimeframe tuple
+        :param data: Tuple containing the DataFrame and the datetime it was cached
+        """
+        if self.__rpc:
+            self.__rpc.send_msg(
+                {
+                    'type': RPCMessageType.ANALYZED_DF,
+                    'data': {
+                        'key': pair_key,
+                        'value': dataframe
+                    }
+                }
+            )
+
+    def add_external_df(
+        self,
+        pair: str,
+        timeframe: str,
+        dataframe: DataFrame,
+        candle_type: CandleType,
+        producer_name: str = "default"
+    ) -> None:
+        """
+        Add the pair data to this class from an external source.
+
+        :param pair: pair to get the data for
+        :param timeframe: Timeframe to get data for
+        :param candle_type: Any of the enum CandleType (must match trading mode!)
+        """
+        pair_key = (pair, timeframe, candle_type)
+
+        if producer_name not in self.__producer_pairs_df:
+            self.__producer_pairs_df[producer_name] = {}
+
+        self.__producer_pairs_df[producer_name][pair_key] = (dataframe, datetime.now(timezone.utc))
+
+    def get_external_df(
+        self,
+        pair: str,
+        timeframe: str,
+        candle_type: CandleType,
+        producer_name: str = "default"
+    ) -> Tuple[DataFrame, datetime]:
+        """
+        Get the pair data from the external sources. Will wait if the policy is
+        set to, and data is not available.
+
+        :param pair: pair to get the data for
+        :param timeframe: Timeframe to get data for
+        :param candle_type: Any of the enum CandleType (must match trading mode!)
+        """
+        pair_key = (pair, timeframe, candle_type)
+
+        if producer_name not in self.__producer_pairs_df:
+            # We don't have this data yet, return empty DataFrame and datetime (01-01-1970)
+            return (DataFrame(), datetime.fromtimestamp(0, tz=timezone.utc))
+
+        if pair_key not in self.__producer_pairs_df:
+            # We don't have this data yet, return empty DataFrame and datetime (01-01-1970)
+            return (DataFrame(), datetime.fromtimestamp(0, tz=timezone.utc))
```

Comment:
I'll not allow "waiting" in freqtrade core bot.

you can wait in the strategy - but waiting means "the world stops" - which means no open trades will be processed. If the strategy does that - it's the user's fault. 
If the bot does that, then its' a bug.

Reply:
It only needs to wait one time. What other choice does it have? It has no idea what key names to fill the DF with if it has never seen a DF before. 

I thought we discussed this a bit in discord already. The idea is that it waits to get its first full DF from the producer. It then caches an empty DF with those column names so that it can feed that to the strat on occasion where it cant get anything from the producer.

Is there an issue with this logic?

Reply:
from dataprovider, you'll get an empty dataframe, or whatever the leader provides you.

Freqtrade will **not wait** for the leader to be ready. This is non-negocible - anything that tries to introduce this will be treated as a bug.
Waiting would mean that no open trades are processed - so (not on exchange) stoplosses will be missed, ROI will be missed, ...

An additional problem that would arise by guaranteeing columns in a leader dataframe:
If we were to wait for the first message so we can fill the dataframe with proper columns - then freqtrade would also take responsibility to ensure all future leader dataframes follow the same structure, reject dataframes that don't (maybe the leader's dataframe generation is not stable?).
That's however not possible - as it might be fine for the strategy if some columns change as long as the key column remains there.

it's problems that freqtrade cannot properly handle - not without preventing a ton of alternative usecases - as there's too many decisions in place, decisions that the strategy can take based on user preference.

Reply:
So just to be clear, you suggest that the follower strategy must accommodate the possibility of an empty dataframe in all of it's various callbacks and populate_* functions?

Reply:
Please read the (resolved) comments we had on this PR. This is now turning into a "wast of time" discussion.

we've had this discussion - and the architecture we've decided on is that the follower will push the data into the dataprovider - and strategies interested in it will pull it from there - as part of the strategy.
This will allow for the most flexibility (the strategy decides how the data should be merged / combined) - how empty or missing data should be treated - as we could never account for all needs as part of core code.

Reply:
To be frank: there are 193 comments on this PR and the majority (~90%, are out dated). My goal for driving this PR is to help the community. Please refrain from telling me I'm wasting anyone's time with this PR or this discussion. 

We did discuss this on discord and we did come to the conclusion that I was suggesting in this thread. Somewhere along the line, stuff changed. Perhaps I missed it - but I'm trying to keep up and I am doing my best to ensure this PR fits the general community needs. 

This is an important point on the PR (indicated by the numerous discord convos we have had it on it already) and yes it may need to be discussed more than once. 

The answer to my question is "yes the user will need to catch an empty data frame throughout their strategy in all various functions."

Fine.

Reply:
The alternative would be to "stop the world" - halting processing of all open trades until a leader submits a dataframe we deem valid.

This can mean stopping for 5 seconds - 1 minute - or 1h (if it's a 1h timeframe - and the last dataframe was for whatever reason invalid).

This automatically implies that the other 49dataframes we received (For the other 49 pairs in the pairlist) will not be acted on - because we got a problem on one single pair - so it'll impact operations of not just open trades - but of everything.

> The answer to my question is "yes the user will need to catch an empty data frame throughout their strategy in all various functions."

This is overstated - and sounds way more complex than it'll really be.
The way i think this should be used is as follows:

* populate_indicators pulls the "external dataframe"
* `populate_indicators` does "something" with this dataframe - returning it (or merged to the main dataframe) to the strategy - at which point it's a regular, "internal", analyzed dataframe.
* `populate_indicators()` will handle the empty dataframe, either by calculating the missing pieces locally, - or providing an empty shell dataframe, with all 0 or NAN values.
  We can provide helpers for this (probably as a future, seperate PR) - but it'll be difficult to cover all cases with that - but we give people with the knowledge the possibility to do it ANY way they like. 
* other methods can now use the same method they use now - to either rely on the `dataframe` parameter - or `dp.get_analyzed_dataframe` where necessary.

Reply:
I will spend some time considering various types of helper functions and how they can be used to help. I agree that it is for a separate PR in the future. For now, like you say, users have the tools to handle an empty dataframe in their strategy.